File: data/code_pairs/fortran/396.f90
! saxpy.f90
subroutine saxpy(n, sa, sx, incx, sy, incy)
    real :: sx(*), sy(*), sa
    integer :: n, incx, incy, i, ix, iy, m, mp1

    if (n <= 0) return
    if (sa == 0.0) return

    if (incx == 1 .and. incy == 1) then
        m = mod(n, 4)
        if (m /= 0) then
            do i = 1, m
                sy(i) = sy(i) + sa * sx(i)
            end do
            if (n < 4) return
        end if
        do i = m+1, n, 4
            sy(i) = sy(i) + sa * sx(i)
            sy(i+1) = sy(i+1) + sa * sx(i+1)
            sy(i+2) = sy(i+2) + sa * sx(i+2)
            sy(i+3) = sy(i+3) + sa * sx(i+3)
        end do
    else
        ix = 1
        iy = 1
        if (incx < 0) ix = (-n + 1) * incx + 1
        if (incy < 0) iy = (-n + 1) * incy + 1
        do i = 1, n
            sy(iy) = sy(iy) + sa * sx(ix)
            ix = ix + incx
            iy = iy + incy
        end do
    endif
end subroutine saxpy

program test_saxpy
    external saxpy
    integer :: n, incx, incy
    real :: sa
    real, allocatable :: sx(:), sy(:)

    n = 5
    sa = 2.0
    incx = 1
    incy = 1
    allocate(sx(n), sy(n))

    sx = (/1.0, 2.0, 3.0, 4.0, 5.0/)
    sy = (/5.0, 4.0, 3.0, 2.0, 1.0/)

    call saxpy(n, sa, sx, incx, sy, incy)

    print *, "SY after SAXPY:"
    print *, sy

    deallocate(sx, sy)
end program test_saxpy

File: data/code_pairs/cpp/396.cpp
#include <iostream>
#include <cmath>
#include <cassert>

void saxpy(int n, float sa, float* sx, int incx, float* sy, int incy) {
    if (n <= 0) return;
    if (sa == 0.0f) return;

    if (incx == 1 && incy == 1) {
        for (int i = 0; i < n; i += 4) {
            sy[i] += sa * sx[i];
            if (i + 1 < n) sy[i + 1] += sa * sx[i + 1];
            if (i + 2 < n) sy[i + 2] += sa * sx[i + 2];
            if (i + 3 < n) sy[i + 3] += sa * sx[i + 3];
        }
    } else {
        int ix = 0;
        int iy = 0;
        if (incx < 0) ix = (n - 1) * incx;
        if (incy < 0) iy = (n - 1) * incy;
        for (int i = 0; i < n; i++) {
            sy[iy] += sa * sx[ix];
            ix += incx;
            iy += incy;
        }
    }
}

void testSaxpy() {
    const int n = 5;
    float sa = 2.0f;
    float sx[n] = {1, 2, 3, 4, 5};
    float sy[n] = {5, 4, 3, 2, 1};
    int incx = 1, incy = 1;

    saxpy(n, sa, sx, incx, sy, incy);

    float expected[n] = {7, 8, 9, 10, 11};
    for (int i = 0; i < n; i++) {
        assert(std::abs(sy[i] - expected[i]) < 1e-5);
    }

    std::cout << "All tests passed." << std::endl;
}

int main() {
    testSaxpy();
    return 0;
}